R/cache-disk.R

```diff
@@ -0,0 +1,533 @@
+      files <- files[!is.na(files$size), ]
+
+      # 1. Remove any files where the age exceeds max age.
+      timediff <- as.numeric(Sys.time() - files[["mtime"]], units = "secs")
+      rm_idx <- timediff > private$max_age
+      if (any(rm_idx)) {
+        private$log(paste0("prune max_age: Removing ", paste(files$name[rm_idx], collapse = ", ")))
+      }
+      file.remove(files$name[rm_idx])
+
+      # Remove rows of files that were deleted.
+      files <- files[!rm_idx, ]
+
+      # Sort files by priority, according to eviction policy.
+      if (private$evict == "lru") {
+        files <- files[order(files[["atime"]], decreasing = TRUE), ]
+      } else if (private$evict == "fifo") {
+        files <- files[order(files[["mtime"]], decreasing = TRUE), ]
+      } else {
+        stop('Unknown eviction policy "', private$evict, '"')
+      }
+
+      # 2. Remove files if there are too many.
+      if (nrow(files) > private$max_n) {
+        rm_idx <- seq_len(nrow(files)) > private$max_n
+        if (any(rm_idx)) {
+          private$log(paste0("prune max_n: Removing ", paste(files$name[rm_idx], collapse = ", ")))
+        }
+        file.remove(files$name[rm_idx])
+      }
+
+      # 3. Remove files if cache is too large.
+      if (sum(files$size) > private$max_size) {
+        cum_size <- cumsum(files$size)
+        rm_idx <- cum_size > private$max_size
+        if (any(rm_idx)) {
+          private$log(paste0("prune max_size: Removing ", paste(files$name[rm_idx], collapse = ", ")))
+        }
+        file.remove(files$name[rm_idx])
+      }
+      invisible(self)
+    },
+
+    size = function() {
+      self$is_destroyed(throw = TRUE)
+      length(dir(private$dir, "\\.rds$"))
+    },
+
+    destroy = function() {
+      if (self$is_destroyed()) {
+        return(invisible(self))
+      }
+
+      private$log(paste0("destroy: Removing ", private$dir))
+      # First create a sentinel file so that other processes sharing this
+      # cache know that the cache is to be destroyed. This is needed because
+      # the recursive unlink is not atomic: another process can add a file to
+      # the directory after unlink starts removing files but before it removes
+      # the directory, and when that happens, the directory removal will fail.
+      file.create(file.path(private$dir, "__destroyed__"))
+      unlink(private$dir, recursive = TRUE)
+      private$destroyed <- TRUE
+      invisible(self)
+    },
+
+    is_destroyed = function(throw = FALSE) {
+      if (!dirExists(private$dir) ||
+          file.exists(file.path(private$dir, "__destroyed__")))
+      {
+        # It's possible for another process to destroy a shared cache directory
+        private$destroyed <- TRUE
+      }
+
+      if (throw) {
+        if (private$destroyed) {
+          stop("Attempted to use cache which has been destroyed:\n  ", private$dir)
+        }
+
+      } else {
+        private$destroyed
+      }
+    },
+
+    finalize = function() {
```

Comment:
Let's not use the finalizer... scary.

Reply:
I left it in there, but now it always defaults to FALSE. I was going to remove it and write a section in the docs about how to use a finalizer to clean up the disk, but I decided it would be simpler to leave this, and only change the default.